### PR TITLE
[FE] Feature: General Management pages

### PR DESCRIPTION
**Pull request by ShinCoder**
### Task

- https://graduation-project-internship.atlassian.net/browse/CI-31

### Features

- Add General Management pages UI + functionality 
    - Consist of list view, detail view, create, update and delete + bulk delete.
    - Include: 
        - District management
        - Ward management
        - Panel type management
        - Report type management

### Screenshots

image
image
image

### CODE REVIEW

File: apps/management-app/src/components/Authenticated/CDO/WardsListView/index.tsx [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+  const handleDeleteWards = useCallback(async () => {
+    if (data) {
+      try {
+        await deleteWards(
+          selectedRow.map((e) => (typeof e === 'number' ? e : parseInt(e))),
+        ).unwrap();
+        showSuccess('Wards deleted');
+      } catch (error) {
+        showError(
+          isApiErrorResponse(error) ? error.data?.message : 'Unknown error',
```
Comment (starhuy1903) [inline apps/management-app/src/components/Authenticated/CDO/WardsListView/index.tsx:54]:
```suggestion
          isApiErrorResponse(error) ? error.data?.message : 'Something went wrong',
```

File: apps/management-app/src/components/Authenticated/CDO/WardsListView/index.tsx [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+            <DeleteIcon sx={{ color: (theme) => theme.palette.common.white }} />
+          </Fab>
+        </>
+      }
+    >
+      <CustomDataGrid
+        rows={rows}
+        columns={columns}
+        checkboxSelection={true}
+        onRowSelectionModelChange={(selected) => setSelectedRow(selected)}
```
Comment (starhuy1903) [inline apps/management-app/src/components/Authenticated/CDO/WardsListView/index.tsx:163]:
```suggestion
        onRowSelectionModelChange={setSelectedRow}
```

File: apps/management-app/src/components/Authenticated/CDO/WardsDetail/index.tsx [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+  useUpdateWardMutation,
+} from '@/store/api/generalManagementApiSlice';
+import { isApiErrorResponse } from '@/store/api/helper';
+import { showError, showSuccess } from '@/utils/toast';
+import FormInputSkeleton from '../FormInputSkeleton';
+import StaticActionBar from '../StaticActionBar';
+
+interface FormData {
+  name: string;
+  district_id: number;
```
Comment (starhuy1903) [inline apps/management-app/src/components/Authenticated/CDO/WardsDetail/index.tsx:25]:
```suggestion
  districtId: number;
```

File: apps/management-app/src/components/Authenticated/CDO/WardsDetail/index.tsx [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    formState: { isDirty, isValid, isLoading },
+    reset,
+  } = useForm<FormData>({
+    resolver: yupResolver(schema),
+    mode: 'onChange',
+    defaultValues: async () => {
+      const ward = await getWard(parseInt(id!), true).unwrap();
+      return ward
+        ? { name: ward.name, district_id: ward.district_id }
+        : { name: '', district_id: 0 };
```
Comment (starhuy1903) [inline apps/management-app/src/components/Authenticated/CDO/WardsDetail/index.tsx:63]: should use try catch

File: apps/management-app/src/components/Authenticated/CDO/WardsDetail/index.tsx [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+            startIcon={<ArrowBackIcon />}
+            onClick={() => navigate(-1)}
+          >
+            Back
+          </Button>
+          <Box sx={{ display: 'inline-flex', columnGap: '16px' }}>
+            <Button
+              variant="outlined"
+              disabled={!isDirty}
+              onClick={() => reset()}
```
Comment (starhuy1903) [inline apps/management-app/src/components/Authenticated/CDO/WardsDetail/index.tsx:114]:
```suggestion
              onClick={reset}
```

File: apps/management-app/src/components/Authenticated/CDO/WardsDetail/index.tsx [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    mode: 'onChange',
+    defaultValues: async () => {
+      const ward = await getWard(parseInt(id!), true).unwrap();
+      return ward
+        ? { name: ward.name, district_id: ward.district_id }
+        : { name: '', district_id: 0 };
+    },
+  });
+
+  const [updateWard] = useUpdateWardMutation();
```
Comment (starhuy1903) [inline apps/management-app/src/components/Authenticated/CDO/WardsDetail/index.tsx:67]: should handle loading status, disable button, form input, ...

File: apps/management-app/src/components/Authenticated/CDO/WardsDetail/index.tsx [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+          <>
+            <ControlledTextField control={control} name="name" label="Name" />
+            <ControlledSelect
+              control={control}
+              name="district_id"
+              label="District"
+              options={districts!.data!.map((e) => ({
+                value: e.id.toString(),
+                label: e.name,
+              }))}
```
Comment (starhuy1903) [inline apps/management-app/src/components/Authenticated/CDO/WardsDetail/index.tsx:161]: districts!.data, what does !. mean?

File: apps/management-app/src/components/Authenticated/CDO/ReportTypesListView/index.tsx [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+            <DeleteIcon sx={{ color: (theme) => theme.palette.common.white }} />
+          </Fab>
+        </>
+      }
+    >
+      <CustomDataGrid
+        rows={rows}
+        columns={columns}
+        checkboxSelection={true}
+        onRowSelectionModelChange={(selected) => setSelectedRow(selected)}
```
Comment (starhuy1903) [inline apps/management-app/src/components/Authenticated/CDO/ReportTypesListView/index.tsx:153]:
```suggestion
        onRowSelectionModelChange={setSelectedRow}
```

File: apps/management-app/src/components/Authenticated/CDO/ReportTypesDetail/index.tsx [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+    if (isError) navigate(-1);
+  }, [isError, navigate]);
+
+  const schema = useMemo(
+    () =>
+      yup.object({
+        name: yup.string().required("Report type's name is required"),
+      }),
+    [],
+  );
```
Comment (starhuy1903) [inline apps/management-app/src/components/Authenticated/CDO/ReportTypesDetail/index.tsx:41]: should define this validation outside of component, on top for example

File: apps/management-app/src/components/Authenticated/CDO/ReportTypesDetail/index.tsx [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+            startIcon={<ArrowBackIcon />}
+            onClick={() => navigate(-1)}
+          >
+            Back
+          </Button>
+          <Box sx={{ display: 'inline-flex', columnGap: '16px' }}>
+            <Button
+              variant="outlined"
+              disabled={!isDirty}
+              onClick={() => reset()}
```
Comment (starhuy1903) [inline apps/management-app/src/components/Authenticated/CDO/ReportTypesDetail/index.tsx:104]:
```suggestion
              onClick={reset}
```

File: apps/management-app/src/components/Authenticated/CDO/PanelTypesListView/index.tsx [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+            <DeleteIcon sx={{ color: (theme) => theme.palette.common.white }} />
+          </Fab>
+        </>
+      }
+    >
+      <CustomDataGrid
+        rows={rows}
+        columns={columns}
+        checkboxSelection={true}
+        onRowSelectionModelChange={(selected) => setSelectedRow(selected)}
```
Comment (starhuy1903) [inline apps/management-app/src/components/Authenticated/CDO/PanelTypesListView/index.tsx:153]:
```suggestion
        onRowSelectionModelChange={setSelectedRow}
```

File: apps/management-app/src/components/Authenticated/CDO/PanelTypesDetail/index.tsx [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+      return panelType ? { name: panelType.name } : { name: '' };
+    },
+  });
+
+  const [updatePanelType] = useUpdatePanelTypeMutation();
+
+  const handleUpdatePanelType = useCallback(
+    handleSubmit(async (data: FormData) => {
+      try {
+        await updatePanelType({ id: parseInt(id!), data }).unwrap();
```
Comment (starhuy1903) [inline apps/management-app/src/components/Authenticated/CDO/PanelTypesDetail/index.tsx:62]: id! -> what does it mean?

File: apps/management-app/src/store/api/generalManagementApiSlice.ts [hunk 13]
```diff
@@ -0,0 +1,269 @@
...
+export const generalManagementApiSlice = apiSlice.injectEndpoints({
+  endpoints: (build) => ({
+    getDistricts: build.query<
+      GetListResult<District>,
+      { page?: number; limit?: number }
+    >({
+      query: (arg) => ({
+        url: `/districts`,
+        params: {
+          page: arg.page || undefined,
```
Comment (starhuy1903) [inline apps/management-app/src/store/api/generalManagementApiSlice.ts:28]:
```suggestion
          page: arg.page,
```

File: apps/management-app/src/store/api/generalManagementApiSlice.ts [hunk 14]
```diff
@@ -0,0 +1,269 @@
...
+        method: 'DELETE',
+        body: { ids: arg },
+      }),
+    }),
+    getPanelTypes: build.query<
+      GetListResult<PanelType>,
+      { page?: number; limit?: number }
+    >({
+      query: (arg) => ({
+        url: `/panelTypes`,
```
Comment (starhuy1903) [inline apps/management-app/src/store/api/generalManagementApiSlice.ts:93]:
```suggestion
        url: `/panel-types`,
```
BE also should change like this

File: apps/management-app/src/components/Authenticated/CDO/WardsListView/index.tsx (hunk 2 above)
Comment (ShinCoder) [reply apps/management-app/src/components/Authenticated/CDO/WardsListView/index.tsx:163]: Row selection removed

File: apps/management-app/src/components/Authenticated/CDO/WardsListView/index.tsx (hunk 1 above)
Comment (ShinCoder) [reply apps/management-app/src/components/Authenticated/CDO/WardsListView/index.tsx:54]: Updated

File: apps/management-app/src/components/Authenticated/CDO/WardsDetail/index.tsx (hunk 3 above)
Comment (ShinCoder) [reply apps/management-app/src/components/Authenticated/CDO/WardsDetail/index.tsx:25]: Updated

File: apps/management-app/src/components/Authenticated/CDO/WardsDetail/index.tsx (hunk 5 above)
Comment (ShinCoder) [reply apps/management-app/src/components/Authenticated/CDO/WardsDetail/index.tsx:114]: Updated

File: apps/management-app/src/components/Authenticated/CDO/WardsDetail/index.tsx (hunk 7 above)
Comment (ShinCoder) [reply apps/management-app/src/components/Authenticated/CDO/WardsDetail/index.tsx:161]: Non-null assertion, changed. Now using districts value to determine render state instead.

File: apps/management-app/src/components/Authenticated/CDO/ReportTypesListView/index.tsx (hunk 8 above)
Comment (ShinCoder) [reply apps/management-app/src/components/Authenticated/CDO/ReportTypesListView/index.tsx:153]: Row selection removed

File: apps/management-app/src/components/Authenticated/CDO/ReportTypesDetail/index.tsx (hunk 9 above)
Comment (ShinCoder) [reply apps/management-app/src/components/Authenticated/CDO/ReportTypesDetail/index.tsx:41]: The schema has component scope and should be wrapped inside its component.

File: apps/management-app/src/components/Authenticated/CDO/PanelTypesDetail/index.tsx (hunk 12 above)
Comment (ShinCoder) [reply apps/management-app/src/components/Authenticated/CDO/PanelTypesDetail/index.tsx:62]: Non-null assertion, the page would never be reached if the id params was empty, typescript doesn't know this.

File: apps/management-app/src/store/api/generalManagementApiSlice.ts (hunk 13 above)
Comment (ShinCoder) [reply apps/management-app/src/store/api/generalManagementApiSlice.ts:28]: Updated

File: apps/management-app/src/store/api/generalManagementApiSlice.ts (hunk 14 above)
Comment (ShinCoder) [reply apps/management-app/src/store/api/generalManagementApiSlice.ts:93]: All endpoints serve as placeholders and will be updated when integrating APIs

File: apps/management-app/src/components/Authenticated/CDO/ReportTypesDetail/index.tsx (hunk 10 above)
Comment (ShinCoder) [reply apps/management-app/src/components/Authenticated/CDO/ReportTypesDetail/index.tsx:104]: Unable to resolve, cause type mismatch.

File: apps/management-app/src/components/Authenticated/CDO/PanelTypesListView/index.tsx (hunk 11 above)
Comment (ShinCoder) [reply apps/management-app/src/components/Authenticated/CDO/PanelTypesListView/index.tsx:153]: Row selection removed

File: apps/management-app/src/components/Authenticated/CDO/ReportTypesDetail/index.tsx (hunk 9 above)
Comment (ShinCoder) [reply apps/management-app/src/components/Authenticated/CDO/ReportTypesDetail/index.tsx:41]: Updated

File: apps/management-app/src/components/Authenticated/CDO/WardsDetail/index.tsx (hunk 4 above)
Comment (ShinCoder) [reply apps/management-app/src/components/Authenticated/CDO/WardsDetail/index.tsx:63]: Updated

File: apps/management-app/src/components/Authenticated/CDO/WardsDetail/index.tsx (hunk 6 above)
Comment (ShinCoder) [reply apps/management-app/src/components/Authenticated/CDO/WardsDetail/index.tsx:67]: Updated

File: apps/management-app/src/components/Authenticated/CDO/AdsTypesDetail/index.tsx [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+    },
+  });
+
+  const [updateAdsType] = useUpdateAdsTypeMutation();
+
+  const handleUpdateAdsType = handleSubmit((data: FormData) => {
+    dispatch(
+      showModal(ModalKey.GENERAL, {
+        headerText: `Apply changes ?`,
+        onModalClose: () => null,
```
Comment (starhuy1903) [inline apps/management-app/src/components/Authenticated/CDO/AdsTypesDetail/index.tsx:69]: we have handled close in `src/components/Modals/index.tsx`

Comment (starhuy1903) [reply apps/management-app/src/components/Authenticated/CDO/AdsTypesDetail/index.tsx:69]: remove it `onModalClose: () => null,`

Comment (ShinCoder) [reply apps/management-app/src/components/Authenticated/CDO/AdsTypesDetail/index.tsx]: Updated

**Review by starhuy1903 — APPROVED** [review]

**Pull request merged**